### PR TITLE
Include result of install.sh in error logs

### PR DESCRIPTION
**Pull request by lidavidm**
Makes #321 easier to debug.

### CODE REVIEW

**Comment by Janzert** [comment]
errors is chopped up right after this section in order to keep the size of it under 3kb, with 1kb coming from the start of errors and 2kb from the end. My guess is that the most useful part of the output is likely to be the end of stderr, unfortunately there is probably a decent chance with this system of that getting removed. The separation messages, "end of install.sh stderr, stdout follows", etc., are also likely to disappear.

The best fix would probably be to keep them all separate until the trimming is done, trim them each separately (maybe with more weight to stderr output over stdout) and then combining with the separation messages. Also the limit should be fine to increase to say 10kb without causing any issue and give a better chance of keeping the important bits.

**Comment by lidavidm** [comment]
@Janzert good idea, I've implemented your suggestion.

File: apiserver/worker/compiler.py [hunk 1]
```diff
@@ -680,63 +680,139 @@ def get_run_lang(submission_dir):
                     if line[0] == '#':
                         return line[1:-1]
 
-def compile_anything(bot_dir, installTimeLimit=600, timelimit=600, max_error_len = 3072):
+
+INSTALL_ERROR_START = "Possible errors running install.sh. stdout output as follows:"
+INSTALL_ERROR_MID = "End of install.sh stdout. stderr as follows:"
+INSTALL_ERROR_END = "End of install.sh stdout."
```
Comment (Janzert) [inline apiserver/worker/compiler.py]: INSTALL_ERROR_END isn't put at the end of the install.sh stdout so instead should probably be something like "End of install.sh output."

File: apiserver/worker/compiler.py [hunk 2]
```diff
@@ -680,63 +680,139 @@ def get_run_lang(submission_dir):
...
-                    end_errors.append(errors[cur_error])
-                    cur_error -= 1
-                    length += len(errors[cur_error])
-                end_errors.reverse()
-                errors = first_errors + end_errors
-
-            return detected_language.name, errors
-    else:
-        return "Unknown", errors
+    if not detected_language or install_errors:
```
Comment (Janzert) [inline apiserver/worker/compiler.py]: I think bailing out here if anything was written to stderr in install.sh is too eager, some programs simply log informational messages to stderr. Currently the compilation is still attempted and only fails if the actual compilation step fails.

File: apiserver/worker/compiler.py [hunk 3]
```diff
@@ -680,63 +680,139 @@ def get_run_lang(submission_dir):
...
+    install_errors = install_errors or []
+    language_detection_errors = language_detection_errors or []
+    compile_errors = compile_errors or []
+
+    all_errors = language_detection_errors + compile_errors
+    result = []
+    if install_errors:
+        all_errors = install_stdout + install_errors
+
+    if sum(len(line) for line in all_errors) <= max_error_len:
```
Comment (Janzert) [inline apiserver/worker/compiler.py]: This isn't including the lengths of langauage_detection and compile errors. I think language detection should always be short but compile errors can be quite long.

File: apiserver/worker/compiler.py [hunk 4]
```diff
@@ -680,63 +680,139 @@ def get_run_lang(submission_dir):
...
+        if install_errors:
```
Comment (Janzert) [inline apiserver/worker/compiler.py]: Even if there was nothing on stderr from install.sh, the stdout should probably still be included.

File: apiserver/worker/compiler.py [hunk 5]
```diff
@@ -680,63 +680,139 @@ def get_run_lang(submission_dir):
...
+                end_errors.append(source[current])
+                length += len(source[current])
+                current -= 1
+            result.extend(reversed(end_errors))
+
+        return length, result
+
+
+    remaining_length = max_error_len
+    if install_errors:
```
Comment (Janzert) [inline apiserver/worker/compiler.py]: Same as above, probably want install.sh stdout even when no stderr output.

File: apiserver/worker/compiler.py (hunk 2 above)
Comment (Janzert) [reply apiserver/worker/compiler.py]: Maybe checking the exit code from install.sh would work, but I'd be a little worried about even doing that given that we haven't up till now.

File: apiserver/worker/compiler.py (hunk 3 above)
Comment (Janzert) [reply apiserver/worker/compiler.py]: Oops, misread the code here. It is including language detection and compile errors in all_errors.

**Comment by lidavidm** [comment]
Thanks for the review, I've addressed those comments.

**Pull request merged**